tests/Microsoft.Identity.Test.Unit/AppTokenProviderTests.cs

```diff
@@ -158,5 +159,60 @@ private AppTokenProviderResult GetAppTokenProviderResult(string differentScopesF
             return token;
         }
 
+        [TestMethod]
+        public async Task MultiThreadedAppTokenProviderCallHasOnlyOneTokenEndpointRequestAsync()
+        {
+            string differentScopesForAt = string.Empty;
+            int totalThreads = 10; // Number of threads to run concurrently
+            int identityProviderHits = 0;
+            int cacheHits = 0;
+            int threadCount = totalThreads;
+
+            var app = ConfidentialClientApplicationBuilder.Create(TestConstants.ClientId)
+                .WithAuthority("https://login.microsoftonline.com/tid")
+                .WithAppTokenProvider((AppTokenProviderParameters parameters) =>
+                {
+                    return Task.FromResult(GetAppTokenProviderResult(differentScopesForAt));
+                })
+                .BuildConcrete();
+
+#pragma warning disable VSTHRD101
+            ParallelLoopResult result = Parallel.For(0, totalThreads, async (i) =>
+            {
+                try
+                {
+                    var authResult = await app.AcquireTokenForClient(TestConstants.s_scope)
+                                                    .ExecuteAsync(new CancellationToken()).ConfigureAwait(false);
+
+                    if (authResult.AuthenticationResultMetadata.TokenSource == TokenSource.IdentityProvider)
+                    {
+                        // Increment identity hits count
+                        Interlocked.Increment(ref identityProviderHits);
+                    }
+                    else
+                    {
+                        // Increment cache hits count
+                        Interlocked.Increment(ref cacheHits);
+                    }
+                }
+                finally
+                {
+                    Interlocked.Decrement(ref threadCount);
+                }
+            });
+
+            while (threadCount != 0)
+            {
+                Thread.Sleep(100);
```

Comment:
As per https://www.oreilly.com/library/view/concurrency-in-c/9781492054498/ - "The moment you use Thread, you have a legacy application". Modern apps should only use Task. 

Consider rewriting this to make use of `Task.WhenAll` https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.whenall?view=net-7.0

Also, a failing test should not wait forever, so pls make sure to add a timeout (e.g. via CancellationToken)_